book/scheduling-and-threading.md

```diff
@@ -312,256 +327,241 @@ situation---some devices have more [CPU cores][cores] than others, or are more
+reason we've been adding async tasks in this chapter!] because the whole point
+of using this API is to keep the website responsive to the user while
+network requests are going on.
+
+Our approach will be to start a thread, run some code on it that does the
+request and gets a response, then schedule a `Task` to send the response back
+to the script. Starting a thread is easy: define a function that is the "main"
+function of the thread, then start the thread with the function passed as the
+`target` argument. When the main function exits, the thread will automatically
+die.
+
+In this example:
 ``` {.python expected=False}
-# This is the rendering event loop we want.
+def run:
+    count = 0
+    while count < 100:
+        print("Thread")
+        count += 1
+
+thread = threading.Thread(target=run)
+thread.start()
 while True:
-    while there_is_enough_time():
-        run_a_task_from_a_task_queue()
-    run_rendering_pipeline()
+    print("Browser")
 ```
 
-The way the loop is written implies that the rendering pipeline is its own task.
-And it's "ideal" because separating rendering into its own task will allow us
-to optimize it and spread it across multiple threads.^[Here
-`run_rendering_pipeline` is made to look like it's all on the main event loop,
-but all the parts of it that are after layout are invisible to web page
-authors, so we'll be able to optimize them later on into a second event loop
-on another thread.]
+a stream of lines with the words "Thread" and "Browser" will be
+interspersed---according to the CPU scheduling algorithm of the computer and
+the Python runtime---until the thread has printed 100 times, after which
+"Browser" will continue printing forever.
 
-But right now, the rendering pipeline in our browser is not a task at
-all---it's hidden in various subroutines of other tasks in the event loop.
+Here is the code for `XMLHttpRequest_send` (the new `is_async` parameter
+indicates an async request). In this case, `run_load` is the thread main
+function, and after the line that says "return out" executes, the thread will
+die.^[Note that for async requests, the return statement is meaningless;
+it's only there for the sync version.]
 
-``` {.python expected=False}
-# This is what our browser currently does.
-while True:
-    run_a_task_from_a_task_queue() # Might do some rendering.
+``` {.python}
+class JSContext:
+    def XMLHttpRequest_send(
+        self, method, url, body, is_async, handle):
+        full_url = resolve_url(url, self.tab.url)
+        if not self.tab.allowed_request(full_url):
+            raise Exception("Cross-origin XHR blocked by CSP")
+
+        def run_load():
+            headers, out = request(
+                full_url, self.tab.url, payload=body)
+            handle_local = handle
+            if url_origin(full_url) != url_origin(self.tab.url):
+                raise Exception(
+                    "Cross-origin XHR request not allowed")
```

Comment:
I think we can do this before starting the thread, right?

Reply:
Done